### ISSUE TITLE
[WASI] Failed to find a pre-opened file descriptor

### ISSUE THREAD

**Issue by XX** [wasmerio/wasmer#2079]
I wrote a simple wasmer-wasi example that reads directory entries but it always fails.

[`wasi_fs_example/src/lib.rs`](https://github.com/XX/wasmer-wasi-fs/blob/master/wasi_fs_example/src/lib.rs):
```rust
#[no_mangle]
pub fn start() {
    std::fs::read_dir("/").unwrap();
}
```

[`runner/src/main.rs`](https://github.com/XX/wasmer-wasi-fs/blob/master/runner/src/main.rs):
```rust
use wasmer::{Instance, Module, Store};
use wasmer_wasi::WasiState;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let wasm = std::fs::read("target/wasm32-wasi/debug/wasi_fs_example.wasm")?;

    let store = Store::default();
    let module = Module::new(&store, wasm)?;

    let mut wasi_env = WasiState::new("wasi_fs_example")
        .preopen_dir("target")?
        .finalize()?;

    let import_object = wasi_env.import_object(&module)?;
    let instance = Instance::new(&module, &import_object)?;

    let start = instance.exports.get_function("start")?;
    start.call(&[])?;

    Ok(())
}
```

This produce the error while running:

```sh
cargo build -p wasi_fs_example --target wasm32-wasi
cargo run -p runner
```
```
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: Custom { kind: Other, error: "failed to find a pre-opened file descriptor through which \"/\" could be opened" }', wasi_fs_example/src/lib.rs:6:10
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
Full source code of the example is located [here](https://github.com/XX/wasmer-wasi-fs). Why doesn't it work?

**Comment by XX** [comment]
Currently in Rust we can really only create WASI binaries, not libraries. WASI only works for the duration of a main function that needs to be there. Calling the wasi functions from anywhere else segfaults because libpreopen isn't initialized.

So to summarize, how to fix this currently:

    Do not have `lib.crate-type` at all
    Have the file be `src/main.rs`, not `src/lib.rs`
    Have `#![no_main]` at the top of the `src/main.rs`
    Use `RUSTFLAGS="-Z wasi-exec-model=reactor" cargo +nightly build --target wasm32-wasi`
    Make sure that you call `_initialize` before anything else
    
https://github.com/WebAssembly/WASI/issues/24
https://github.com/rust-lang/rust/pull/79997

**Comment by XX** [comment]
In my case:

[`wasi_fs_example/src/main.rs`](https://github.com/XX/wasmer-wasi-fs/blob/master/wasi_fs_example/src/main.rs):
```rust
#![no_main]

#[no_mangle]
pub extern "C" fn start() {
    std::fs::read_dir("/").unwrap();
}
```

[`runner/src/main.rs`](https://github.com/XX/wasmer-wasi-fs/blob/master/runner/src/main.rs):
```rust
...
let start = instance.exports.get_function("_initialize")?;
start.call(&[])?;

let start = instance.exports.get_function("start")?;
start.call(&[])?;
...
```

```sh
RUSTFLAGS="-Z wasi-exec-model=reactor" cargo +nightly build -p wasi_fs_example --target wasm32-wasi
```

**Comment by MarkMcCaskey** [comment]
Interesting thanks for the info!

One thing I immediately noticed is the name of your start function. Wasmer CLI calls `_start`, not `start`. I'm not familiar with the Rust flags you're using to compile (`RUSTFLAGS="-Z wasi-exec-model=reactor"`) but if it's not being set up correctly with initialize (as you point out) it won't work, so that's where I'd look first

**Comment by XX** [comment]
@MarkMcCaskey Oh, consider that there is the name `foo` instead of `start`. It is a library, not a binary with `_start` and `main`. And for the library, WASI works correctly only with the `#![no_main]`/`RUSTFLAGS="-Z wasi-exec-model=reactor"` hack.

**Comment by syrusakbary** [comment]
We believe this issue was fixed by #2941. We need to create a test verifying this

**Comment by ptitSeb** [comment]
After some fideling with the tests, I realized it's already there: `wasi_sees_virtual_root.rs` is exactly this test.
Ticket can be closed!
